base/math.jl

```diff
@@ -102,6 +102,77 @@ macro horner(x, p...)
     return Expr(:block, :(local t = $(esc(x))), ex, :r)
 end
 
+"""
+    evalpoly(x, p::Tuple)
+
+Evaluate the polynomial ``\\sum_k p[k] x^{k-1}`` for the coefficients `p[1]`, `p[2]`, ...;
+that is, the coefficients are given in ascending order by power of `x`. This function
+generates efficient code using Horner's method with loops unrolled at compile time.
+
+# Example
+```jldoctest
+julia> evalpoly(2, (1, 2, 3))
+17
+```
+"""
+function evalpoly(x, p::Tuple)
+    if @generated
+        N = length(p.parameters)
+        ex = :(p[end])
+        for i in N-1:-1:1
+            ex = :(muladd(x, $ex, p[$i]))
+        end
+        ex
+    else
+        N = length(p)
+        ex = p[end]
+        for i in N-1:-1:1
+            ex = muladd(x, ex, p[i])
+        end
+        ex
+    end
+end
+
+
+"""
+    evalpoly(z::Complex, p::Tuple)
+
+Evaluate the polynomial ``\\sum_k p[k] z^{k-1}`` for the coefficients `p[1]`, `p[2]`, ...;
+that is, the coefficients are given in ascending order by power of `z`. This function
+generates efficient code using a Goertzel-like algorithm specialized for complex arguments
+with loops unrolled at compile time.
```

Comment:
Would be good to add the Knuth citation somewhere.

Reply:
Done.